react-app/src/modules/account/account.js

```diff
@@ -0,0 +1,100 @@
+import scrypt from 'scrypt-async';
+import nacl from 'tweetnacl';
+import sha256 from 'fast-sha256';
+import blake from 'blakejs';
+
+nacl.util = require('tweetnacl-util');
+
+let _username, _pk, _sk, _tz, _offset;
+
+export function clearAccount() {
+  _username = _pk = _sk = _tz = _offset = undefined;
+}
+
+export function setKeyPair(pk, sk) {
+  _pk = nacl.util.decodeBase64(pk);
+  _sk = nacl.util.decodeBase64(sk);
+}
+
+function _generateNonce(publicKeyA, publicKeyB) {
+  let state = blake.blake2bInit(nacl.box.nonceLength, null);
+	blake.blake2bUpdate(state, publicKeyA);
+  blake.blake2bUpdate(state, publicKeyB);
+  return blake.blake2bFinal(state);
+}
+
+export function constructor(username, password) {
+  return new Promise(resolve => {
+    scrypt(sha256(nacl.util.decodeUTF8(password)), sha256(nacl.util.decodeUTF8(username)), {
+      N: 1 << 20,
+      r: 8,
+      p: 1,
+      dkLen: 32,
+      encoding: 'binary'
+    }, function(derivedKey) {
+      const keyPair = nacl.box.keyPair.fromSecretKey(derivedKey);
+      _username = username;
+      _pk = keyPair.publicKey;
+      _sk = keyPair.secretKey;
+      resolve(null);
+    });
+  });
+}
+
+export function bytes() {
+  return JSON.stringify({
+    username: _username,
+    pk: _pk,
+    tz: _tz,
+    offset: _offset
+  });
+}
+
+export function encrypt(plainText) {
+  const messageToEncrypt = nacl.util.decodeUTF8(plainText);
+  let output =
+    new Uint8Array(nacl.box.publicKeyLength + nacl.box.overheadLength + messageToEncrypt.length);
```

Comment:
This length is not necessarily enough. You don't know that the ciphertext length is the same as the plainText length. You should either the ciphertext before you create this output.

Reply:
nacl.box() returns an encrypted message nacl.box.overheadLength longer than the original one. Nevertheless, I will rewrite this lines because they may generate confusions.

Reply:
Fair. You're not wrong, but it's also dependent on the underlying crypto. Doing it the other way is "safer", I guess. But also the chances of the underlying crypto to change are almost none.

For the future, this would be a good point to push back, don't hesitate to start a discussion, we will all learn from it :)